File: escomvoting/src/model/response/BallotDTO.ts
export interface BallotDTO {
  id: string
  candidateId: string
  candidateName: string
  voterGroup: string        // STUDENT | PROFESSOR | ADMIN
  nullifier: string         // hex64
  rPrime: string            // uncompressed EC point hex (130 chars)
  sPrime: string            // hex64
  ePrime: string            // hex64
  publicKeyForRole: string  // uncompressed EC point hex for this ballot's group
  submittedAt: string       // ISO-8601
}

File: escomvoting-backend/src/main/java/com/github/escom/escomvoting/model/dto/BallotDTO.java
package com.github.escom.escomvoting.model.dto;

import com.github.escom.escomvoting.model.entity.Ballot;
import com.github.escom.escomvoting.model.entity.Election;

import java.time.Instant;
import java.util.UUID;

public record BallotDTO(
        UUID id,
        UUID candidateId,
        String candidateName,
        String voterGroup,
        String nullifier,
        String rPrime,
        String sPrime,
        String ePrime,
        String publicKeyForRole,
        Instant submittedAt
) {
    public static BallotDTO from(Ballot b, Election election) {
        String pubKey = election.getPublicKeyForRole(b.getVoterGroup());
        return new BallotDTO(
                b.getId(),
                b.getCandidate().getId(),
                b.getCandidate().getName(),
                b.getVoterGroup().name(),
                b.getNullifier(),
                b.getRPrime(),
                b.getSPrime(),
                b.getEPrime(),
                pubKey,
                b.getSubmittedAt()
        );
    }
}
